File: SingleNumber/SingleNumberII/SingleNumberII.py
class Solution(object):
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        # ones = 0
        # twos = 0
        # threes = 0
        # for i in nums:
        #     twos |= ones & i
        #     ones ^= i
        #     threes = ones & twos
        #     ones &= ~threes
        #     twos &= ~threes
        # return ones
        ones = 0
        twos = 0
        for i in nums:
            ones = (ones ^ i) & ~twos
            twos = (twos ^ i) & ~ones
        return ones

File: SingleNumber/SingleNumberII/SingleNumberII.cpp
class Solution {
public:
    int singleNumber(vector<int>& nums) {
        vector<int> tmp(32);
        for(auto num : nums) {
        	for(int j = 31; j >= 0; j--) {
        		tmp[j] += num & 1;
        		num >>= 1;
        		if(!num)
        			break;
        	}
        }
        int result = 0;
        for(int j = 31; j >=0 ;j--) {
        	int t = tmp[j] % 3;
        	if(t)
        		result += 1 << (31 - j);
        }
        return result;
    }
};